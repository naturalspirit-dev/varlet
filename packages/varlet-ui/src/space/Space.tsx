import { computed, defineComponent } from 'vue'
import { props, sizeValidator } from './props'
import './space.less'
import { toPxNum, isPx } from '../utils/elements'
import { SpaceSize } from '../../types'
import { isNumber, isString } from '../utils/shared'

const sizefilter = (size: SpaceSize): [string, string?] => {
  if (!sizeValidator(size)) return ['normal']
  if (isString(size)) {
    if (!size.match(/\d/g)) return [size]
    return [`${toPxNum(size) / 2}px`, `${toPxNum(size) / 2}px`]
  }
  if (isNumber(size)) return [`${toPxNum(size) / 2}px`, `${toPxNum(size) / 2}px`]
  if (Array.isArray(size)) {
    return [`${toPxNum(size[0]) / 2}px`, `${toPxNum(size[1]) / 2}px`]
  }
  return ['normal']
}

export default defineComponent({
  name: 'VarSpace',
  props,
  setup(props, { slots }) {
    const itemSize = computed(() => {
      return sizefilter(props.size)
    })
    console.log(itemSize)
    const [marginTop, marginRight] = itemSize.value as any
    const children = slots.default?.() ?? []
    const lastIndex = children.length - 1
    const renderSpace = children.map((child, index) => {
      const style = {}
      if (props.direction === 'column') {
        if (index === lastIndex) {
          Object.assign(style, {
            margin: 0,
          })
        } else {
          Object.assign(style, {
            marginTop: 0,
            marginBottom: isPx(marginTop) ? (props.justify === 'space-around' ? 0 : marginTop) : '',
            marginRight: 0,
            marginLeft: 0,
          })
        }
      }
      if (props.direction === 'row') {
        if (index === lastIndex) {
          Object.assign(style, {
            marginTop: isPx(marginTop) ? marginTop : '',
            marginBottom: isPx(marginTop) ? marginTop : '',
            marginRight: 0,
            marginLeft: 0,
          })
        } else {
          Object.assign(style, {
            marginTop: isPx(marginTop) ? marginTop : '',
            marginBottom: isPx(marginTop) ? marginTop : '',
            marginRight: isPx(marginRight) ? (props.justify === 'space-around' ? 0 : marginRight) : '',
            marginLeft: 0,
          })
        }
      }

      return (
        <div class={['var-space__item', !isPx(marginTop) ? `var-space__item--${marginTop}` : '']} style={style}>
          {child}
        </div>
      )
    })
    return () => {
      itemSize
      return (
        <>
          <div
            class={[
              'var-space',
              props.inline ? 'var-space--inline' : '',
              `var-space--${props.direction}`,
              `var-space--justifyContent__${props.justify}`,
              props.align ? `var-space--${props.align}` : '',
              `var-space--${props.wrap ? 'warp' : 'nowarp'}`,
            ]}
          >
            {renderSpace}
          </div>
        </>
      )
    }
  },
})
