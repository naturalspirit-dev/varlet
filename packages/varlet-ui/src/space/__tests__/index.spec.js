import example from '../example'
import VarSpace from '../Space'
import Space from '..'
import { mount } from '@vue/test-utils'
import { createApp, h } from 'vue'

test('test space example', () => {
    const wrapper = mount(example)
    expect(wrapper.html()).toMatchSnapshot()
    wrapper.unmount()
})

test('test space plugin', () => {
    const app = createApp({}).use(Space)
    expect(app.component(Space.name)).toBeTruthy()
})

test('test space props', async () => {
    const wrapper = mount(VarSpace)

    await wrapper.setProps({ direction: 'row' })
    expect(wrapper.html()).toMatchSnapshot()


    await wrapper.setProps({ direction: 'column' })
    expect(wrapper.html()).toMatchSnapshot()

    await wrapper.setProps({ size: 'mini' })
    expect(wrapper.html()).toMatchSnapshot()

    await wrapper.setProps({ size: 'small' })
    expect(wrapper.html()).toMatchSnapshot()

    await wrapper.setProps({ size: 'normal' })
    expect(wrapper.html()).toMatchSnapshot()

    await wrapper.setProps({ size: 'large' })
    expect(wrapper.html()).toMatchSnapshot()

    await wrapper.setProps({ size: '10px' })
    expect(wrapper.html()).toMatchSnapshot()

    await wrapper.setProps({ size: '10rem' })
    expect(wrapper.html()).toMatchSnapshot()

    await wrapper.setProps({ size: '[10px,20px]' })
    expect(wrapper.html()).toMatchSnapshot()

    await wrapper.setProps({ wrap: true })
    expect(wrapper.html()).toMatchSnapshot()

    await wrapper.setProps({ wrap: false })
    expect(wrapper.html()).toMatchSnapshot()

    await wrapper.setProps({ wrap: false })
    expect(wrapper.html()).toMatchSnapshot()

    await wrapper.setProps({ justify: 'start' })
    expect(wrapper.html()).toMatchSnapshot()

    await wrapper.setProps({ justify: 'end' })
    expect(wrapper.html()).toMatchSnapshot()

    await wrapper.setProps({ justify: 'center' })
    expect(wrapper.html()).toMatchSnapshot()

    await wrapper.setProps({ justify: 'space-between' })
    expect(wrapper.html()).toMatchSnapshot()

    await wrapper.setProps({ justify: 'space-around' })
    expect(wrapper.html()).toMatchSnapshot()

    await wrapper.setProps({ align: 'start' })
    expect(wrapper.html()).toMatchSnapshot()

    await wrapper.setProps({ align: 'center' })
    expect(wrapper.html()).toMatchSnapshot()

    await wrapper.setProps({ align: 'end' })
    expect(wrapper.html()).toMatchSnapshot()

    await wrapper.setProps({ align: 'stretch' })
    expect(wrapper.html()).toMatchSnapshot()

    await wrapper.setProps({ align: 'baseline' })
    expect(wrapper.html()).toMatchSnapshot()

    await wrapper.setProps({ align: 'initial' })
    expect(wrapper.html()).toMatchSnapshot()

    await wrapper.setProps({ align: 'inherit' })
    expect(wrapper.html()).toMatchSnapshot()

    await wrapper.setProps({ inline: true })
    expect(wrapper.html()).toMatchSnapshot()

    await wrapper.setProps({ inline: false })
    expect(wrapper.html()).toMatchSnapshot()

    wrapper.unmount()
})

test('test div in space', async () => {
    const wrapper = mount({
        data: () => ({
            size: 'mini'
        }),
        render() {
            return h(
                VarSpace,
                {
                    size: this.size
                },
                () => [h('div'), h('div'), h('div'), h('div')]
            )
        }
    })
    expect(wrapper.html()).toMatchSnapshot()

    await wrapper.setData({
        size: ['10px', '20px']
    })
    expect(wrapper.html()).toMatchSnapshot()
    wrapper.unmount()
})