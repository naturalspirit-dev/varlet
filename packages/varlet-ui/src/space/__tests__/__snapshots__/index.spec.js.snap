// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test div in space 1`] = `
"<div class=\\"var-space var--box\\" style=\\"flex-direction: row; justify-content: start; flex-wrap: wrap; margin-top: -2px;\\">
  <div style=\\"margin: 2px 2px 2px 0px;\\">
    <div></div>
  </div>
  <div style=\\"margin: 2px 2px;\\">
    <div></div>
  </div>
  <div style=\\"margin: 2px 2px;\\">
    <div></div>
  </div>
  <div style=\\"margin: 2px 0px 2px 2px;\\">
    <div></div>
  </div>
</div>"
`;

exports[`test div in space 2`] = `
"<div class=\\"var-space var--box\\" style=\\"flex-direction: row; justify-content: start; flex-wrap: wrap; margin-top: -5px;\\">
  <div style=\\"margin: 5px 10px 5px 0px;\\">
    <div></div>
  </div>
  <div style=\\"margin: 5px 10px;\\">
    <div></div>
  </div>
  <div style=\\"margin: 5px 10px;\\">
    <div></div>
  </div>
  <div style=\\"margin: 5px 0px 5px 10px;\\">
    <div></div>
  </div>
</div>"
`;

exports[`test div in space 3`] = `
"<div class=\\"var-space var--box var-space--inline\\" style=\\"flex-direction: column; justify-content: start; flex-wrap: wrap; margin-top: -5px; margin-bottom: -5px;\\">
  <div style=\\"margin: 0px 0px 10px 0px;\\">
    <var-button>Button1</var-button>
  </div>
  <div style=\\"margin: 0px 0px 10px 0px;\\">
    <var-button>Button2</var-button>
  </div>
  <div style=\\"margin: 0px;\\">
    <var-button>Button3</var-button>
  </div>
</div>"
`;

exports[`test div in space 4`] = `
"<div class=\\"var-space var--box\\" style=\\"flex-direction: column; justify-content: start; flex-wrap: wrap; margin-top: -5px; margin-bottom: -5px;\\">
  <div style=\\"margin: 0px 0px 10px 0px;\\">
    <var-button>Button1</var-button>
  </div>
  <div style=\\"margin: 0px 0px 10px 0px;\\">
    <var-button>Button2</var-button>
  </div>
  <div style=\\"margin: 0px;\\">
    <var-button>Button3</var-button>
  </div>
</div>"
`;

exports[`test space example 1`] = `
"<div class=\\"app-type\\">基本用法</div>
<div class=\\"var-space var--box\\" style=\\"flex-direction: row; justify-content: start; flex-wrap: wrap; margin-top: -4px;\\">
  <div style=\\"margin: 4px 6px 4px 0px;\\"><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--default var-elevation--1\\">
      <!--v-if-->
      <div class=\\"var-button__content\\">Button1</div>
    </button></div>
  <div style=\\"margin: 4px 6px;\\"><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--default var-elevation--1\\">
      <!--v-if-->
      <div class=\\"var-button__content\\">Button2</div>
    </button></div>
  <div style=\\"margin: 4px 0px 4px 6px;\\"><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--default var-elevation--1\\">
      <!--v-if-->
      <div class=\\"var-button__content\\">Button3</div>
    </button></div>
</div>
<div class=\\"app-type\\">垂直</div>
<div class=\\"var-space var--box\\" style=\\"flex-direction: column; justify-content: start; flex-wrap: wrap;\\">
  <div style=\\"margin: 0px 0px 12px 0px;\\"><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--default var-elevation--1\\">
      <!--v-if-->
      <div class=\\"var-button__content\\">Button1</div>
    </button></div>
  <div style=\\"margin: 0px 0px 12px 0px;\\"><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--default var-elevation--1\\">
      <!--v-if-->
      <div class=\\"var-button__content\\">Button2</div>
    </button></div>
  <div style=\\"margin: 0px;\\"><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--default var-elevation--1\\">
      <!--v-if-->
      <div class=\\"var-button__content\\">Button3</div>
    </button></div>
</div>
<div class=\\"app-type\\">间隙</div>
<div class=\\"var-space var--box\\" style=\\"flex-direction: row; justify-content: start; flex-wrap: wrap; margin-top: -15px;\\">
  <div style=\\"margin: 15px 5px 15px 0px;\\"><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--default var-elevation--1\\">
      <!--v-if-->
      <div class=\\"var-button__content\\">Button1</div>
    </button></div>
  <div style=\\"margin: 15px 5px;\\"><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--default var-elevation--1\\">
      <!--v-if-->
      <div class=\\"var-button__content\\">Button2</div>
    </button></div>
  <div style=\\"margin: 15px 0px 15px 5px;\\"><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--default var-elevation--1\\">
      <!--v-if-->
      <div class=\\"var-button__content\\">Button3</div>
    </button></div>
</div>
<div class=\\"app-type\\">靠右</div>
<div class=\\"var-space var--box\\" style=\\"flex-direction: row; justify-content: end; flex-wrap: wrap; margin-top: -4px;\\">
  <div style=\\"margin: 4px 6px 4px 0px;\\"><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--default var-elevation--1\\">
      <!--v-if-->
      <div class=\\"var-button__content\\">Button1</div>
    </button></div>
  <div style=\\"margin: 4px 0px 4px 6px;\\"><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--default var-elevation--1\\">
      <!--v-if-->
      <div class=\\"var-button__content\\">Button2</div>
    </button></div>
</div>"
`;

exports[`test space props 1`] = `"<div class=\\"var-space var--box\\" style=\\"flex-direction: row; justify-content: start; flex-wrap: wrap; margin-top: -4px;\\"></div>"`;

exports[`test space props 2`] = `"<div class=\\"var-space var--box\\" style=\\"flex-direction: column; justify-content: start; flex-wrap: wrap; margin-top: -4px;\\"></div>"`;

exports[`test space props 3`] = `"<div class=\\"var-space var--box\\" style=\\"flex-direction: column; justify-content: start; flex-wrap: wrap; margin-top: -4px;\\"></div>"`;

exports[`test space props 4`] = `"<div class=\\"var-space var--box\\" style=\\"flex-direction: column; justify-content: start; flex-wrap: wrap; margin-top: -4px;\\"></div>"`;

exports[`test space props 5`] = `"<div class=\\"var-space var--box\\" style=\\"flex-direction: column; justify-content: start; flex-wrap: wrap; margin-top: -4px;\\"></div>"`;

exports[`test space props 6`] = `"<div class=\\"var-space var--box\\" style=\\"flex-direction: column; justify-content: start; flex-wrap: wrap; margin-top: -4px;\\"></div>"`;

exports[`test space props 7`] = `"<div class=\\"var-space var--box\\" style=\\"flex-direction: column; justify-content: start; flex-wrap: wrap; margin-top: -4px;\\"></div>"`;

exports[`test space props 8`] = `"<div class=\\"var-space var--box\\" style=\\"flex-direction: column; justify-content: start; flex-wrap: wrap; margin-top: -4px;\\"></div>"`;

exports[`test space props 9`] = `"<div class=\\"var-space var--box\\" style=\\"flex-direction: column; justify-content: start; flex-wrap: wrap; margin-top: -4px;\\"></div>"`;

exports[`test space props 10`] = `"<div class=\\"var-space var--box\\" style=\\"flex-direction: column; justify-content: start; flex-wrap: wrap; margin-top: -4px;\\"></div>"`;

exports[`test space props 11`] = `"<div class=\\"var-space var--box\\" style=\\"flex-direction: column; justify-content: start; flex-wrap: nowrap; margin-top: -4px;\\"></div>"`;

exports[`test space props 12`] = `"<div class=\\"var-space var--box\\" style=\\"flex-direction: column; justify-content: start; flex-wrap: nowrap; margin-top: -4px;\\"></div>"`;

exports[`test space props 13`] = `"<div class=\\"var-space var--box\\" style=\\"flex-direction: column; justify-content: start; flex-wrap: nowrap; margin-top: -4px;\\"></div>"`;

exports[`test space props 14`] = `"<div class=\\"var-space var--box\\" style=\\"flex-direction: column; justify-content: end; flex-wrap: nowrap; margin-top: -4px;\\"></div>"`;

exports[`test space props 15`] = `"<div class=\\"var-space var--box\\" style=\\"flex-direction: column; justify-content: center; flex-wrap: nowrap; margin-top: -4px;\\"></div>"`;

exports[`test space props 16`] = `"<div class=\\"var-space var--box\\" style=\\"flex-direction: column; justify-content: space-between; flex-wrap: nowrap; margin-top: -4px;\\"></div>"`;

exports[`test space props 17`] = `"<div class=\\"var-space var--box\\" style=\\"flex-direction: column; justify-content: space-around; flex-wrap: nowrap; margin-top: -4px;\\"></div>"`;

exports[`test space props 18`] = `"<div class=\\"var-space var--box\\" style=\\"flex-direction: column; justify-content: space-around; flex-wrap: nowrap; margin-top: -4px; align-items: start;\\"></div>"`;

exports[`test space props 19`] = `"<div class=\\"var-space var--box\\" style=\\"flex-direction: column; justify-content: space-around; flex-wrap: nowrap; margin-top: -4px; align-items: center;\\"></div>"`;

exports[`test space props 20`] = `"<div class=\\"var-space var--box\\" style=\\"flex-direction: column; justify-content: space-around; flex-wrap: nowrap; margin-top: -4px; align-items: end;\\"></div>"`;

exports[`test space props 21`] = `"<div class=\\"var-space var--box\\" style=\\"flex-direction: column; justify-content: space-around; flex-wrap: nowrap; margin-top: -4px; align-items: stretch;\\"></div>"`;

exports[`test space props 22`] = `"<div class=\\"var-space var--box\\" style=\\"flex-direction: column; justify-content: space-around; flex-wrap: nowrap; margin-top: -4px; align-items: baseline;\\"></div>"`;

exports[`test space props 23`] = `"<div class=\\"var-space var--box\\" style=\\"flex-direction: column; justify-content: space-around; flex-wrap: nowrap; margin-top: -4px; align-items: initial;\\"></div>"`;

exports[`test space props 24`] = `"<div class=\\"var-space var--box\\" style=\\"flex-direction: column; justify-content: space-around; flex-wrap: nowrap; margin-top: -4px; align-items: inherit;\\"></div>"`;

exports[`test space props 25`] = `"<div class=\\"var-space var--box var-space--inline\\" style=\\"flex-direction: column; justify-content: space-around; flex-wrap: nowrap; margin-top: -4px; align-items: inherit;\\"></div>"`;

exports[`test space props 26`] = `"<div class=\\"var-space var--box\\" style=\\"flex-direction: column; justify-content: space-around; flex-wrap: nowrap; margin-top: -4px; align-items: inherit;\\"></div>"`;
